jsonschema-core.xml

```diff
@@ -288,6 +278,36 @@
                         zeros) are insignificant.
                     </t>
                 </section>
+
+                <section title="Non-JSON Instances">
+                    <t>
+                        It is possible to use JSON Schema with a superset of the JSON Schema data model,
+                        where an instance may be outside any of the six JSON data types.
+                    </t>
+                    <t>
+                        In this case, annotations still apply; but validation keywords will not be useful,
+                        as they will always pass or always fail.
+                    </t>
+                    <t>
+                        A custom vocabulary may define support for a superset of the core data model.
+                        The schema itself may only be expressible in this superset;
+                        for example, to make use of the "const" keyword.
+                    </t>
+                    <t>
+                        New data model types may be defined as either supersets of an existing type
```

Comment:
Please clarify if that means that existing keywords can be extended, or if new keywords must be created to extend functionality. This paragraph almost implies that a vocabulary might extend the existing 'type' keyword to accept more values than number, integer, boolean, string, null, object and array.

Reply:
> This paragraph almost implies that a vocabulary might extend the existing 'type' keyword to accept more values than number, integer, boolean, string, null, object and array.

This is exactly the intent.

Reply:
FWIW I think that's terrible. It means that the implementation of every core keyword now needs to check if another vocabulary has extended that keyword. For example: when the validator vocabulary evaluates "type": "my_new_type"`, it can no longer generate an error "unrecognized type 'my_new_type'" but instead.. what?  silently ignore it? call the other vocabulary's evaluation code to handle it? Is it going to know what vocabulary that is?

Reply:
Vocabulary behavior aside, it makes sense to me that if I'm going to produce an implementation that's a superset of JSON Schema, that implementation will allow schemas like `{ type: "binary" }`.

I'll take another look and see if I can address this though.

I feel like this has come up once or twice before, or is this just déjà vu?

Reply:
> it makes sense to me that if I'm going to produce an implementation that's a superset of JSON Schema...

I agree.  Perhaps 'type' is the only keyword where it might make sense to expand the syntax of an existing keyword. If so, we can handle this thusly:
* update the documentation for 'type' to indicate it should explicitly ignore any unrecognized type names, to allow for new vocabularies to add their own 'type' keyword (and indicate that those vocabularies should also ignore all the core types and any other unrecognized values) - which would allow vocabularies to share the 'type' keyword between them
* explicitly document that vocabularies cannot use the same keyword name as a core vocabulary (with the single exception of 'type')
* leave as undefined behaviour what happens when two custom vocabularies choose to use the same keyword (some vocabularies may choose to error on a collision; others may wish to have overlapping behaviour as with 'type').

Reply:
> I feel like this has come up once or twice before, or is this just déjà vu?

I've brought this up wrt the OAS 3.1 vocabulary definition. I know the spec says vocabularies are allowed to modify keywords, but this is incredibly problematic to implement in a system that allows users to provide their own vocabulary implementations. My implementation doesn't support adding/modifying keyword semantics other than what can be constrained by using an alternate meta-schema (such as disallowing the array form of the `type` keyword).

However, you can have a vocabulary that defines a keyword called `type` with the new semantics. That means that if your vocabulary provides a different `type` keyword, and you include the validation vocabulary, you now have two keywords with the name `type`. Therefore, you also have to provide a custom validation vocabulary definition that copies the standard one except leaves out `type`.

Adding a custom `type` keyword is something that can be supported, but you can't also use the standard validation vocabulary. If the new type requires modifications to other keywords, those would also conflict with the standard vocabularies they appear in.

Reply:
@awwright in the previous PR you wrote:

> So the specification defining "type": "binary" I think is out.

Which I took to mean that we would ***NOT*** be making `type` extensible.  I realize now that that is not quite what that sentence says, but I agree wih @karenetheridge and @jdesrosiers that we really shouldn't do that.  There is no way to express it in the vocabulary construction.

@jdesrosiers's efforts looking into similarly problematic attempts to square differing definitions of `exclusiveMinimum` and `exclusiveMaximum` across drafts led me to abandon that attempt, and helped me make the case to OAS to drop compatibility so we didn't need it.

> Adding a custom `type` keyword is something that can be supported, but you can't also use the standard validation vocabulary. If the new type requires modifications to other keywords, those would also conflict with the standard vocabularies they appear in.

> I know the spec says vocabularies are allowed to modify keywords

We should look at that.  What it should say is that compatible overlapping definitions are allowed, or something like that.  If a keyword has a fixed `enum` set of values in one vocabulary, and other values in another, I would argue that that is _not_ compatible (for the same reason that we abandoned the `exclusiveMinimum` stuff).

------------

I think it is fine to say that extra keywords can be added.  For example, some sort of `isBinary` assertion (since `contentMediaType` is never an assertion).

To the extent that this interacts with `type`, it just removes any implicit assumption that the absence of `type` is the same as `"type": ["null", "boolean", "number", "array", "object"]`.  If `type` is absent and you have a parsed instance, no further type checking is done as an assertion.

-------------

I think there's an open question about disambiguating keywords that could make sense for binary values but don't automatically, like `minLength`.  But the more I think about it the more I lean towards saying that you should add your own keywords, `minBinaryLength`, for example.

In practice, a modular implementation will dispatch the instance to keyword handlers based on what the vocabulary describes.  Each module should be able to make a decision on handling without coordinating with other modules.  What that really means is that a keyword can't fail an assertion in one module and pass in another, _and expect the result to be a pass_.  If any handler fails, the result is a failure.  Just like adding keywords to a schema object, constraints can only be added, never removed.

Does this make sense?

Reply:
This all sounds good.  I also found this in the spec (https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.7.2.3): "Vocabularies do not support specifically declaring different value sets for keywords." which at the moment explicitly rules out adding additional valid values to the 'type' keyword, but we could always carve out an exception for it as we do for 'format' (see one paragraph up in the spec.  Adding new keywords e.g. isBinaryType would certainly be less painful though.

Reply:
@karenetheridge we didn't really carve out an exception for `format`.  It's always been defined as having an open-ended set of values and we are forced to continue supporting that for the time being.  No one should ever create a keyword like `format` again.

Reply:
@handrews I well understand the difficult nature of `format` :)  I am saying we should be explicit in the spec (given the questions and different understandings shown earlier in this thread) that it is not generally permitted for a vocabulary to extend the syntax, range of permitted values or other functionality of an existing core keyword.  Especially in the context of extending the defined data model set (the topic of the PR), the spec should give direction that new data types will need vocabularies to define *new* keywords to work with them, rather than extend existing keywords.

Reply:
My read on the above comments are, this PR is off, and the key is...

> To the extent that this interacts with type, it just removes any implicit assumption that the absence of type is the same as "type": ["null", "boolean", "number", "array", "object"]. If type is absent and you have a parsed instance, no further type checking is done as an assertion. - @handrews 

In addition, +1 to what @handrews further said...

> What that really means is that a keyword can't fail an assertion in one module and pass in another, and expect the result to be a pass. If any handler fails, the result is a failure. Just like adding keywords to a schema object, constraints can only be added, never removed.

@awwright do you feel you've misread the issue here?
Do you feel a new PR would make the most sense or do you want to continue to modify this one?